ark/jsonschema/object.ts

```diff
@@ -0,0 +1,248 @@
+						actual: key
+					})
+				}
+			}
+			for (const key in jsonSchema.properties)
+				if (!jsonSchema.required.includes(key)) optionalKeys.push(key)
+		} else {
+			// If 'required' is not present, all keys are optional
+			optionalKeys.push(...Object.keys(jsonSchema.properties))
+		}
+	} else if ("required" in jsonSchema) {
+		ctx.reject({
+			expected: "a valid object JSON Schema",
+			actual:
+				"an object JSON Schema with 'required' array but no 'properties' object"
+		})
+	}
+
+	return {
+		optionalKeys: optionalKeys.map(key => ({
+			key,
+			value: parseJsonSchema(jsonSchema.properties![key]).internal
+		})),
+		requiredKeys: requiredKeys.map(key => ({
+			key,
+			value: parseJsonSchema(jsonSchema.properties![key]).internal
+		}))
+	}
+}
+
+const parseAdditionalProperties = (jsonSchema: JsonSchema.Object) => {
+	if (!("additionalProperties" in jsonSchema)) return
+
+	const properties =
+		jsonSchema.properties ? Object.keys(jsonSchema.properties) : []
+	const patternProperties = Object.keys(jsonSchema.patternProperties ?? {})
+
+	const additionalPropertiesSchema = jsonSchema.additionalProperties
+	if (additionalPropertiesSchema === true) return true
+	if (additionalPropertiesSchema === false) return false
+
+	const schemaDefinedKeys = rootSchema(
+		[...properties]
+			.map(key => ({ unit: key }))
+			.concat(
+				[...patternProperties].map(key => ({
+					domain: "string",
+					pattern: key
+				})) as never
+			)
+	)
+
+	const jsonSchemaObjectAdditionalPropertiesValidator = (
+		data: object,
+		ctx: Traversal
+	) => {
+		Object.keys(data).forEach(key => {
+			if (schemaDefinedKeys.allows(key))
+				// Not an additional property, so don't validate here
+				return
+
+			const additionalPropertyValidator = parseJsonSchema(
+				additionalPropertiesSchema
+			)
+
+			const value = data[key as keyof typeof data]
+			if (!additionalPropertyValidator.allows(value)) {
+				ctx.reject({
+					path: [key],
+					expected: `${additionalPropertyValidator.description}, since ${key} is an additional property.`,
+					actual: printable(value)
+				})
+			}
+		})
+		return !ctx.hasError()
+	}
+	return jsonSchemaObjectAdditionalPropertiesValidator
+}
+
+export const parseObjectJsonSchema: Type<
+	(In: JsonSchema.Object) => Out<Type<object, any>>,
+	any
+> = JsonSchemaScope.ObjectSchema.pipe((jsonSchema, ctx): Type<object> => {
+	const arktypeObjectSchema: Intersection.Schema<object> = {
+		domain: "object"
+	}
+
+	const { requiredKeys, optionalKeys } = parseRequiredAndOptionalKeys(
+		jsonSchema,
+		ctx
+	)
+	arktypeObjectSchema.required = requiredKeys
+	arktypeObjectSchema.optional = optionalKeys
+
+	const patternProperties = parsePatternProperties(jsonSchema)
+	if (patternProperties) arktypeObjectSchema.index = patternProperties
+
+	const potentialPredicates: (Predicate.Schema | undefined)[] = [
+		...parseMinMaxProperties(jsonSchema, ctx),
```

Comment:
man so many crazy object constraints in JSON schema. thanks for implementing all these!